### PR TITLE
Ignore NotConnected error in poll_shutdown()

### LINKED ISSUE #41: Error `NotConnected` when serving a `TlsStream<TcpStream>` when using `hyper`
When serving a `TlsStream<TcpStream>` using a `hyper` server, the stream's `poll_shutdown` method sometimes returns a `io::ErrorKind::NotConnected`. This happens when calling the server using curl with `HTTP/1.1`, but not with `HTTP/1.0`. Also, it happens when calling the server using `openssl s_client` and pressing `ctrl+c` instead of `ctrl+d`.

It seems that when a client doesn't explicitly close the connection, but simply hangs up, the stream ends with an error.

This only happens with `tokio_rustls` and not with `tokio_native_tls`.

Here is a somewhat minimal reproduction of the error:

```rust
use std::{
	fs::File,
	io::BufReader,
	net::{Ipv4Addr, SocketAddrV4},
	sync::Arc,
};

use hyper::{Request, Response};
use hyper_util::rt::TokioIo;
use tokio::net::TcpListener;
use tokio_native_tls::native_tls::Identity;
use tokio_rustls::{rustls::pki_types::CertificateDer, TlsAcceptor};

#[tokio::main]
async fn main() {
	let bind_addr = SocketAddrV4::new(Ipv4Addr::new(127, 0, 0, 1), 8080);
	let listener = TcpListener::bind(bind_addr).await.unwrap();
	let service = hyper::service::service_fn(move |_request: Request<_>| async {
		Result::<_, std::io::Error>::Ok(Response::new("".to_string()))
	});

	println!("Listening on {bind_addr}");
	loop {
		let (stream, _addr) = listener.accept().await.unwrap();
		tokio::spawn(async move {
			let tls_acceptor = get_tls_acceptor();
			// Uncomment to use native_tls
			// let tls_acceptor = get_tls_acceptor_native();

			let tls_stream = tls_acceptor.accept(stream).await.unwrap();
			hyper::server::conn::http1::Builder::new()
				.preserve_header_case(true)
				.title_case_headers(true)
				.serve_connection(TokioIo::new(tls_stream), service)
				.with_upgrades()
				.await
				.unwrap()
		});
	}
}

fn get_tls_acceptor_native() -> tokio_native_tls::TlsAcceptor {
	let key = std::fs::read("certs/key.pem").unwrap();
	let cert = std::fs::read("certs/certs.pem").unwrap();
	let identity = Identity::from_pkcs8(&cert, &key).unwrap();
	let acceptor = tokio_native_tls::native_tls::TlsAcceptor::builder(identity).build().unwrap();
	tokio_native_tls::TlsAcceptor::from(acceptor)
}

fn get_tls_acceptor() -> tokio_rustls::TlsAcceptor {
	let mut key_reader = BufReader::new(File::open("certs/key.pem").unwrap());
	let mut cert_reader = BufReader::new(File::open("certs/certs.pem").unwrap());

	let key = rustls_pemfile::private_key(&mut key_reader).unwrap().unwrap();
	let certs = rustls_pemfile::certs(&mut cert_reader)
		.map(|cert| cert.map(CertificateDer::from))
		.collect::<std::result::Result<_, std::io::Error>>()
		.unwrap();

	let config = tokio_rustls::rustls::ServerConfig::builder()
		.with_no_client_auth()
		.with_single_cert(certs, key)
		.unwrap();

	TlsAcceptor::from(Arc::new(config))
}
``` 

To reproduce using curl:

`curl -k -v https://localhost:8080 --http1.0` gives no error.
`curl -k -v https://localhost:8080 --http1.1` gives a `NotConnected` error.

To reproduce using  `openssl s_client`:

`openssl s_client -connect localhost:8080` then `Ctrl+D` gives no error.
`openssl s_client -connect localhost:8080` then `Ctrl+C` gives a `NotConnected` error.

### PR DESCRIPTION
**Pull request by djc**
Fixes #41.

### CODE REVIEW

**Review by cpu — APPROVED** [review]
This seems reasonable to me based on what you found in https://github.com/rustls/tokio-rustls/issues/41

WDYT about adding some brief context to the commit message?

**Comment by djc** [comment]
Added some context to the commit message.

**Comment by TCROC** [comment]
I ran into this issue as well with reqwest.  This commit definitely _reduces_ the amount of errors that I get, but it doesn't fix it completely.  When testing hitting https://google.com 100 times, I actually start getting different responses than I get with openssl AND I occasionally get this error now:

```
connection error: unexpected end of file
```

I suspect that maybe EOF errors need to be ignored as well?

**Comment by TCROC** [comment]
> > I suspect that maybe EOF errors need to be ignored as well?
> > ..
> > And then still occasionally returns:
> > error sending request for url (https://google.com/): connection error: unexpected end of file
> 
> I think this is being addressed upstream ([hyperium/hyper#3427](https://github.com/hyperium/hyper/issues/3427), [hyperium/h2#743](https://github.com/hyperium/h2/pull/743)) and is separate from the concerns of this branch. See also the Rustls manual's section on [Unexpected EOF](https://docs.rs/rustls/latest/rustls/manual/_03_howto/index.html#unexpected-eof).
> 
> The other two responses seem fine to me. The choice of the server's 301 or 302 response doesn't seem obviously tied to Rustls.

I added h2 as a dependency and pointed it to the git repo:

```
h2 = { git = "https://github.com/hyperium/h2.git" }
```

But I still get the error:

```
Got an error: error sending request for url (https://google.com/): connection error: unexpected end of file
```

**Comment by TCROC** [comment]
These are my reqwest dependencies:

```
reqwest = { version = "0.11", default-features = false, features = [
    "rustls-tls",
] }
tokio = { version = "1", default-features = false }
tokio-rustls = { git = "https://github.com/rustls/tokio-rustls.git", rev = "8c9d6f32b972a3c3882a304f8e5156034805ed43" }
h2 = { git = "https://github.com/hyperium/h2.git" }
```

**Comment by cpu** [comment]
Wouldn't you need to point your `reqwest` dependency at a fork that also takes a patch to use `h2`'s unreleased git branch with the change? Does `cargo tree` reflect what you expect w.r.t the patched h2 dependency? I suspect you ended up with two h2 versions and reqwest using h2 0.3.14

Since this seems unrelated to the change in this PR I think it would be more appropriate to discuss in a separate issue.

**Comment by TCROC** [comment]
> Wouldn't you need to point your `reqwest` dependency at a fork that also takes a patch to use `h2`'s unreleased git branch with the change? Does `cargo tree` reflect what you expect w.r.t the patched h2 dependency? I suspect you ended up with two h2 versions and reqwest using h2 0.3.14
> 
> Since this seems unrelated to the change in this PR I think it would be more appropriate to discuss in a separate issue.

Yep I agree.  I actually wasn't able to test updating h2 alongside this fix as there were compiler errors when I cloned reqwest.  I'm just going to stick to using openssl for now and will circle back around to rustls when reqwest updates with the patches :).

Thanks for the help! :)

**Comment by ctz** [comment]
I went looking for why this doesn't affect OpenSSL -- as of OpenSSL 3 they have the same behaviour as rustls. (Though there is a flag (`SSL_OP_IGNORE_UNEXPECTED_EOF`) that can be set via `SSL_set_options` to turn off that behaviour --but I can't see any rust code anywhere on github that references that option.)

The answer is a bug in rust-openssl that mishandles the error: https://github.com/sfackler/rust-openssl/issues/1803

**Comment by djc** [comment]
@quininer do you want to/have time to take a look at this?

File: src/common/mod.rs [hunk 1]
```diff
@@ -333,7 +333,11 @@ where
 
     fn poll_shutdown(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
         while self.session.wants_write() {
-            ready!(self.write_io(cx))?;
+            match ready!(self.write_io(cx)) {
+                Ok(_) => {}
+                Err(err) if err.kind() == io::ErrorKind::NotConnected => {}
```
Comment (quininer) [inline src/common/mod.rs:338]: maybe it should `break`? otherwise it could loop infinitely.

**Review by quininer — COMMENTED** [review]
Sorry for delay, I just finished my Spring Festival vacation.

A more reasonable approach is to record a state to ensure that `write_io` is no longer called after call `io.shutdown`. but that's fine too.

**Comment by djc** [comment]
> Sorry for delay, I just finished my Spring Festival vacation.

No worries, hope you had a nice vacation!

> A more reasonable approach is to record a state to ensure that `write_io` is no longer called after call `io.shutdown`. but that's fine too.

Ahh, you think this is because `poll_shutdown()` is called twice? (Looked this up, the `std` documentation notes that calling it a second time on Linux will yield `Ok(())` but on macOS it will yield `io::ErrorKind::NotConnected`.)

#41 states it's happening because the peer closed the connection.

**Comment by quininer** [comment]
I'm think about something like this

```rust
fn poll_shutdown(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
    while !self.start_shutdown && self.session.wants_write() {
        ready!(self.write_io(cx))?;
    }

    self.start_shutdown = true;
    Pin::new(&mut self.io).poll_shutdown(cx)
}
```

I believe this is because the signal has been emitted when `io.shutdown` was called, but still returned pending, and when `poll_shutdown` is called for second time, `write_io` will return an error.

**Comment by cpu** [comment]
> Looked this up, the std documentation notes that calling it a second time on Linux will yield Ok(()) but on macOS it will yield io::ErrorKind::NotConnected.
...
> I believe this is because the signal has been emitted when io.shutdown was called, but still returned pending, and when poll_shutdown is called for second time, write_io will return an error.

Interesting! For what it's worth yesterday I spent a little bit of time trying to write a test for this condition. I couldn't manage to get `shutdown` to error despite trying a few different approaches. In hindsight I was testing on Linux and so was likely on a pointless quest.

@Revanee Out of curiosity, in #41 when you were observing this issue were you testing on MacOS?

**Comment by Revanee** [comment]
@cpu I've tested it on Arch Linux with kernel `6.7.0-arch3-1`. The example in the issue should be enough to reproduce the error. Here is my `Cargo.toml` with the dependency versions I used:

```toml
[package]
name = "proxy_test"
version = "0.1.0"
edition = "2021"

[dependencies]
hyper = { version = "1.1.0", features = ["full"] }
hyper-util = { version = "0.1.1", features = ["full"] }
tokio = { version = "1.18.1", features = ["macros", "rt-multi-thread", "sync"] }
tokio-native-tls = "0.3.1"
tokio-rustls = "0.25.0"
rustls-pemfile = "2.0.0"
```

**Comment by djc** [comment]
@Revanee would you be able to test whether your issue is fixed with the proposal from @quininer?

**Comment by Revanee** [comment]
@djc I tested both the changes from the PR, as well as the proposed changes from @quininer. Neither seem to fix the issue for me. In both cases `Pin::new(&mut self.io).poll_shutdown(cx)` returns the error `Err(Os { code: 107, kind: NotConnected, message: "Transport endpoint is not connected" })`.

**Comment by djc** [comment]
Okay, I've pushed a simpler fix, can you test this?

**Comment by djc** [comment]
@Revanee do you have time to test this soon? We'd like to get a new release out soonish with the rustls 0.23 upgrade, and it would be nice to ship a fix for this at the same time.

**Comment by Revanee** [comment]
@djc I tested the new fix and it works! :)

**Comment by djc** [comment]
Thanks! @quininer what do you think of this? It seems like the minimal fix that addresses the observed issue, we can always extend this to more cases if they are encountered in the wild by someone.

**Review by quininer — APPROVED** [review]
LGTM

**Comment by quininer** [comment]
I think this is good, at least it doesn't bring risks.

**Pull request merged**